src/librustc_mir/build/scope.rs

```diff
@@ -695,10 +695,17 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 if let DropKind::Value { .. } = drop_kind {
                     scope.needs_cleanup = true;
                 }
+
                 let region_scope_span = region_scope.span(self.hir.tcx(),
                                                           &self.hir.region_scope_tree);
-                // Attribute scope exit drops to scope's closing brace
-                let scope_end = region_scope_span.with_lo(region_scope_span.hi());
+                // Attribute scope exit drops to scope's closing brace.
+                // Without this check when finding the endpoint, we'll run into an ICE.
```

Comment:
Should this be somehow handled in `end_point`?

Reply:
I think it would make sense to have a check in `end_point` that makes sure it doesn't cause any overflow when doing `self.hi().0 - 1`.

I'm not entirely sure why the MIR borrow checker has any spans where that would be an issue (particularly since the AST borrow checker doesn't need this check). I only noticed this when compiling `rustc_tsan` in the std artifacts compilation step.

Reply:
It _is_ weird. Could you make that change so that any other code calling `end_point` doesn't need to worry about this?

Also, if you could post the ICE that happens in `rustc_tsan` it would be great, as I am intrigued at which piece of code it was triggering this (my guess is that it just made an existing bug apparent).

Reply:
I'm already compiling a change that moves that check into `end_point` as I write this.

The only error it gave me that is left in my scrollback is this, which isn't very useful, sorry about that:

```
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.25.0-dev running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'attempt to subtract with overflow', libsyntax_pos/lib.rs:222:27
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

I didn't bother to look into it further than that when I confirmed it was my change that caused it. I did notice though that of the four instances of the ICE in my scrollback, they occured in different modules: `rustc_lsan`, `rustc_asan`, `compiler_builtins` and `rustc_tsan` - each run was probably with minor adjustments to my code to try work out what was happening.

Reply:
Cool. I'll r+ once that last change is in (and CI has run successfully).